taskchampion/lib/src/replica.rs

```diff
@@ -186,6 +238,272 @@ pub unsafe extern "C" fn tc_replica_new_on_disk(
+                // SAFETY:
+                //  - timestamp is valid (promised by caller)
+                //  - caller will not use timestamp after this call (convention)
+                let timestamp_ruststring = unsafe { TCString::val_from_arg(timestamp) };
+
+                ReplicaOp::Update {
+                    uuid: uuid_ruststring,
+                    property: property_ruststring.into_string().unwrap(),
+                    old_value: Some(old_value_ruststring.into_string().unwrap()),
+                    value: Some(value_ruststring.into_string().unwrap()),
+                    timestamp: timestamp_ruststring
+                        .into_string()
+                        .unwrap()
+                        .parse::<DateTime<Utc>>()
+                        .unwrap(),
+                }
+            }
+            TCReplicaOp {
+                operation_type: TCReplicaOpType::UndoPoint,
+                ..
+            } => ReplicaOp::UndoPoint,
+            TCReplicaOp {
+                operation_type: TCReplicaOpType::Error,
+                ..
+            } => panic!("This shouldn't be possible."),
+        }
+    }
+}
+
+#[ffizz_header::item]
+#[ffizz(order = 901)]
+/// ***** TCReplicaOpList *****
+///
+/// ```c
+/// struct TCReplicaOpList {
+///     struct TCReplicaOp *items;
+///     size_t len;
+///     size_t capacity;
+/// };
+///
+/// typedef struct TCReplicaOpList TCReplicaOpList;
+/// ```
+#[repr(C)]
+pub struct TCReplicaOpList {
+    items: *mut TCReplicaOp,
+    len: usize,
+    capacity: usize,
+}
+
+impl CList for TCReplicaOpList {
+    type Element = TCReplicaOp;
+
+    unsafe fn from_raw_parts(items: *mut Self::Element, len: usize, cap: usize) -> Self {
+        TCReplicaOpList {
+            len,
+            capacity: cap,
+            items,
+        }
+    }
+
+    fn slice(&mut self) -> &mut [Self::Element] {
+        // SAFETY:
+        //  - because we have &mut self, we have read/write access to items[0..len]
+        //  - all items are properly initialized Element's
+        //  - return value lifetime is equal to &mmut self's, so access is exclusive
+        //  - items and len came from Vec, so total size is < isize::MAX
+        unsafe { std::slice::from_raw_parts_mut(self.items, self.len) }
+    }
+
+    fn into_raw_parts(self) -> (*mut Self::Element, usize, usize) {
+        (self.items, self.len, self.capacity)
+    }
+}
+
+impl From<Vec<ReplicaOp>> for TCReplicaOpList {
+    fn from(replica_op_list: Vec<ReplicaOp>) -> TCReplicaOpList {
+        // XXX Remove prints
+        println!("{:#?}", replica_op_list);
+        let tc_replica_op_list: Vec<TCReplicaOp> = replica_op_list.into_iter().map(|op| TCReplicaOp::from(op)).collect();
+        println!("{:#?}", tc_replica_op_list);
+        TCReplicaOpList {
+            items: tc_replica_op_list.as_ptr() as *mut TCReplicaOp,
+            len: tc_replica_op_list.len(),
+            capacity: tc_replica_op_list.capacity(),
+        }
```

Comment:
I think the reason you're getting garbage is that this is creating a new Vec, then embedding a pointer to it in the return value, then freeing the Vec on return from this method. So the resulting TCReplicaOpList is pointing to unallocated memory!

To achieve this, I think you'll need to make the TCReplicaOpList "own" the Vec. You can do that with [`Vec::into_raw_parts`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.into_raw_parts), which both returns the raw parts of the Vec and also "forgets" the original Vec so that it won't be freed on return. You'd need to pair that with `Vec::from_raw_parts` in `tc_replica_op_list_free`, to ensure that memory was freed.

Reply:
That was it!

However, can we really use `Vec::into_raw_parts` when it is a nightly/experimental API? I tried using `Box::leak` instead but I'm not confident that's right. I got a `free(): invalid pointer` error when attempting to free it.

Reply:
Sorry, I forgot about that! There is an equivalent utility function in taskchampion-lib somewhere (vec_into_raw_parts?)